components/scream/src/share/util/scream_tridiag.hpp

```diff
@@ -0,0 +1,564 @@
+  impl::thomas_factorize(team, dl, d, du);
+  team.team_barrier();
+  impl::thomas_solve(team, dl, d, du, X);
+}
+
+template <typename TridiagDiag, typename DataArray>
+KOKKOS_INLINE_FUNCTION
+void thomas (TridiagDiag dl, TridiagDiag d, TridiagDiag du, DataArray X,
+             typename std::enable_if<TridiagDiag::rank == 1>::type* = 0,
+             typename std::enable_if<DataArray::rank == 1>::type* = 0,
+             impl::EnableIfCanUsePointer<TridiagDiag>* = 0,
+             impl::EnableIfCanUsePointer<DataArray>* = 0) {
+  const int nrow = d.extent_int(0);
+  assert( X.extent_int(0) == nrow);
+  assert(dl.extent_int(0) == nrow);
+  assert(du.extent_int(0) == nrow);
+  impl::thomas_a1x1(dl.data(), d.data(), du.data(), X.data(), nrow);
+}
+
+template <typename TridiagDiag, typename DataArray>
+KOKKOS_INLINE_FUNCTION
+void thomas (TridiagDiag dl, TridiagDiag d, TridiagDiag du, DataArray X,
+             typename std::enable_if<TridiagDiag::rank == 1>::type* = 0,
+             typename std::enable_if<DataArray::rank == 2>::type* = 0,
+             impl::EnableIfCanUsePointer<TridiagDiag>* = 0,
+             impl::EnableIfCanUsePointer<DataArray>* = 0) {
+  const int nrow = d.extent_int(0);
+  const int nrhs = X.extent_int(1);
+  assert( X.extent_int(0) == nrow);
+  assert(dl.extent_int(0) == nrow);
+  assert(du.extent_int(0) == nrow);
+  impl::thomas_a1xm(dl.data(), d.data(), du.data(), X.data(), nrow, nrhs);
+}
+
+template <typename TridiagDiag, typename DataArray>
+KOKKOS_INLINE_FUNCTION
+void thomas (TridiagDiag dl, TridiagDiag d, TridiagDiag du, DataArray X,
+             typename std::enable_if<TridiagDiag::rank == 2>::type* = 0,
+             typename std::enable_if<DataArray::rank == 2>::type* = 0,
+             impl::EnableIfCanUsePointer<TridiagDiag>* = 0,
+             impl::EnableIfCanUsePointer<DataArray>* = 0) {
+  const int nrow = d.extent_int(0);
+  const int nrhs = X.extent_int(1);
+  assert(X .extent_int(0) == nrow);
+  assert(dl.extent_int(0) == nrow);
+  assert(du.extent_int(0) == nrow);
+  assert(dl.extent_int(1) == nrhs);
+  assert(d .extent_int(1) == nrhs);
+  assert(du.extent_int(1) == nrhs);
+  impl::thomas_amxm(dl.data(), d.data(), du.data(), X.data(), nrow, nrhs);
+}
+
+// Cyclic reduction at the Kokkos team level. Any (thread, vector)
+// parameterization is intended to work.
+template <typename TeamMember, typename TridiagDiag, typename DataArray>
+KOKKOS_INLINE_FUNCTION
+void cr (const TeamMember& team,
+         TridiagDiag dl, TridiagDiag d, TridiagDiag du, DataArray X,
+         typename std::enable_if<TridiagDiag::rank == 1>::type* = 0,
+         typename std::enable_if<DataArray::rank == 1>::type* = 0) {
+  using Scalar = typename TridiagDiag::non_const_value_type;
+  const int nrow = d.extent_int(0);
+  assert(dl.extent_int(0) == nrow);
+  assert(d. extent_int(0) == nrow);
+  assert(du.extent_int(0) == nrow);
+  assert(X. extent_int(0) == nrow);
+  const int team_id = impl::get_thread_id_within_team(team);
```

Comment:
I think team_id may be mistaken as the id of the team, rather than the id within the team. Perhaps calling it team_thread_id is less confusing?

Reply:
Do you object to an overlap in nomenclature with Kokkos? Otherwise, look at the two multi-RHS impls of `cr` that follow this one. In these, `team_id` and `tid` are separate, and `team_id` makes sense, if one is willing to permit the solver to have its own conception of how h/w threads are gathered into a team. In this first impl of `cr` you're looking at, `tid == team_id`, and I decided to use the second so the loop calculations would be the same as in the other two `cr` impls.

Reply:
It's not a strong objection, really. It's just that when going over the impl, and reading 'team_id', my brain translated that into 'id of the team' (so basically a league rank) rather than 'id of thread (or warp) within the team'. I had to scroll up a couple of times to remind myself that the latter was the meaning, not the former.

Anyhow, the impl details in this file are not likely to be often looked at, so this is not really a concern.